### PR TITLE
Added attribute to store UID ranges

### PR DESCRIPTION
**Pull request by metodej**
- added new attributes to store UID ranges instead of UID max/min
- renamed method used to check and convert GID and now UID ranges

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 1]
```diff
@@ -1085,24 +1085,24 @@ public User getUserFromMessage(PerunSessionImpl sess, String message) throws Int
...
 		//Check every range if it is in correct format and it is valid range
-		for(String minimumOfRange: gidRanges.keySet()) {
+		for(String minimumOfRange: idRanges.keySet()) {
 			//Check not null
-			if(minimumOfRange == null || minimumOfRange.isEmpty()) throw new WrongAttributeValueException(gidRangesAttribute, "Minimum in one of gid ranges is empty!");
-			String maximumOfRange = gidRanges.get(minimumOfRange);
-			if(maximumOfRange == null || maximumOfRange.isEmpty()) throw new WrongAttributeValueException(gidRangesAttribute, "Maximum in one of gid ranges is empty!");
+			if(minimumOfRange == null || minimumOfRange.isEmpty()) throw new WrongAttributeValueException(idRangesAttribute, "Minimum in one of gid ranges is empty!");
+			String maximumOfRange = idRanges.get(minimumOfRange);
+			if(maximumOfRange == null || maximumOfRange.isEmpty()) throw new WrongAttributeValueException(idRangesAttribute, "Maximum in one of gid ranges is empty!");
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: In the text there should be 'one of id ranges'.

File: perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java [hunk 2]
```diff
@@ -1128,7 +1128,7 @@ public User getUserFromMessage(PerunSessionImpl sess, String message) throws Int
 		//Check gid ranges overlaps (there should be no overlaps)
 		Integer lastMaxValue = 0;
 		for(Integer minValue: convertedRanges.keySet().stream().sorted().collect(Collectors.toList())) {
-			if(minValue <= lastMaxValue) throw new WrongAttributeValueException(gidRangesAttribute, "There is an overlap between two gid ranges.");
+			if(minValue <= lastMaxValue) throw new WrongAttributeValueException(idRangesAttribute, "There is an overlap between two gid ranges.");
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/ModulesUtilsBlImpl.java]: Text should be 'between two id ranges'.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		} catch(WrongReferenceAttributeValueException ex) {
+			throw new WrongReferenceAttributeValueException(attribute, ex.getReferenceAttribute());
+		}
+
+	}
+
+	@Override
+	public Attribute fillAttribute(PerunSessionImpl sess, Facility facility, AttributeDefinition attributeDefinition) {
+		return new Attribute(attributeDefinition);
+	}
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java]: I don't think you need to override this method if there is no change in it.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+		//Not suported yet.
+		throw new InternalErrorException("Can't remove value of this virtual attribute this way. " + attributeDefinition);
+	}
+
+	private Attribute getNamespaceUIDRangesAttribute(PerunSessionImpl sess, String uidNamespace) throws InternalErrorException {
+		try {
+			return sess.getPerunBl().getAttributesManagerBl().getAttribute(sess, uidNamespace, A_E_namespaceUIDRanges);
+		} catch(AttributeNotExistsException ex) { throw new ConsistencyErrorException(ex);
+		} catch(WrongAttributeAssignmentException ex) { throw new InternalErrorException(ex);
+		}
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java]: This is not wrong, but I would prefer to have a throw on separate line, please.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	public boolean setAttributeValue(PerunSessionImpl sess, Facility facility, Attribute attribute) throws InternalErrorException, WrongReferenceAttributeValueException {
+		Attribute uidNamespaceAttribute = getUidNamespaceAttribute(sess, facility);
+		if(uidNamespaceAttribute.getValue() == null) throw new WrongReferenceAttributeValueException(attribute, uidNamespaceAttribute, "There is missing UID namespace on the facility.");
+		Attribute namespaceUIDRangesAttribute = getNamespaceUIDRangesAttribute(sess, uidNamespaceAttribute.valueAsString());
+		if(! (attribute.getValue() == null ? namespaceUIDRangesAttribute.getValue() == null : attribute.getValue().equals(namespaceUIDRangesAttribute.getValue()))) {
+			//attribute from param have other vale then physical attribute
+			throw new WrongReferenceAttributeValueException(attribute, namespaceUIDRangesAttribute, "You can't change attribute value of UID Ranges by changing value of virtual attribute.");
+		}
+		return false;
+	}
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java]: Maybe it is problem with gid ranges too, but I am not sure why is this logic here at all. You definitely can't set anything this way (there is no logic for setting in it) so maybe we should skip it at all. Or did I miss something?

**Comment by metodej** [comment]
Thank you for review. I corrected the leftover "gid" in comments of exceptions, I've also added test for one of the modules, the other one just uses one method from ModulesUtils... I've deleted the setAttribute method, I took it from gidRanges attribute but I don't see any reason for it, sorry about that.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class urn_perun_facility_attribute_def_virt_UIDRanges extends FacilityVirtualAttributesModuleAbstract implements FacilityVirtualAttributesModuleImplApi {
+
+	private static final String A_E_namespaceUIDRanges = AttributesManager.NS_ENTITYLESS_ATTR_DEF + ":namespace-UIDRanges";
+	private static final String A_FAC_uidNamespace = AttributesManager.NS_FACILITY_ATTR_DEF + ":uid-namespace";
+
+	@Override
+	public void checkAttributeSemantics(PerunSessionImpl sess, Facility facility, Attribute attribute) throws InternalErrorException, WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
```
Comment (zlamalp) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java:22]: We no longer use ``InternalErrorException`` in the API, please remove it from all the methods.

Comment (metodej) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java]: Okay, removed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+		return strongDependencies;
+	}
+
+	@Override
+	public AttributeDefinition getAttributeDefinition() {
+		AttributeDefinition attr = new AttributeDefinition();
+		attr.setNamespace(AttributesManager.NS_FACILITY_ATTR_VIRT);
+		attr.setFriendlyName("UIDRanges");
+		attr.setDisplayName("UID ranges in set namespace for the Facility");
+		attr.setType(List.class.getName());
```
[the remark this answers is not in the document]
Comment (metodej) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_facility_attribute_def_virt_UIDRanges.java]: It has the correct type now, thank you.

**Pull request merged**